### DISCUSSION TITLE
Round inner & outer corners of transition from tube to wedge

### DISCUSSION THREAD

**Discussion by pimlie** [BelfrySCAD/BOSL2#1460] [Q&A]
Was wondering if BOSL can help with rounding both the inner as outer corners of the transition from a tube to a wedge or a tube to a rect tube?

F.e. given the following shapes, what is the best way to round both the outer as inner corners (marked in red) at height=10?
```openscad
// tube to wedge
tube(h=10, or=18, ir=16);
translate([0, 0, 10])
tube(h=10, or1=18, ir1=16, or2=12, ir2=10);
```
![image](https://github.com/user-attachments/assets/e48c0c4c-be60-41b2-a261-02e448f61983)
```openscad
// rect tube to tube
h1=10;
or1=18; ir1=16;
or2=12; ir2=10;
difference() {
  hull() {
    cube([4*or2, or2, 0.01], center=true);
    translate([0,0,h1]) cylinder(r=or1,h=0.01);
  }
  translate([0,0,-0.01]) 
  hull() {
    cube([4*ir2, ir2, 0.02], center=true);
    translate([0,0,h1+0.01]) cylinder(r=ir1,h=0.02);
  }
}
translate([0,0,h1])
tube(h=10, or=or1, ir=ir1, center=false);
```
![image](https://github.com/user-attachments/assets/991da764-ee03-414b-a5f9-6c29a349b3cd)

Using basic scad you can use minkowski to round the corners for the first shape. Note that the inner corner requires another difference but have omitted here that to keep the example simple. But I'm still wondering if there isn't a more flexible solution?

```openscad
h1=10; h2=10;
or1=18; ir1=16;
or2=12; ir2=10;
radius=0.99;

rotate_extrude(angle=360) {
  intersection() {
    minkowski() {
      polygon([
        [ir1 + radius, 0],
        [ir1 + radius, h1],
        [ir2 + radius, h1 + h2],
        [or2 - radius, h1 + h2],
        [or1 - radius, h1],
        [or1 - radius, 0]
      ]);
      circle(r=radius);
    }

    translate([15, 10])
    rect([12, 20]);
  }
}
```

I guess that [example 21](https://github.com/BelfrySCAD/BOSL2/wiki/turtle3d.scad) from the turtle3d wiki page comes somewhat close to what Im looking for, but that still feels quite complicated to have to use to connect tubes all the time :)

**Comment by pimlie** [comment]
For `tube to wedge` I was able to create this profile using `path_join`'s which you can then rotate_extrude:

![image](https://github.com/user-attachments/assets/a7c6bec6-4441-4def-8bad-025bafaf7785)

The only downside of this profile is that I didn't manage to add full chamfer/rounding support on the bottom/top edges. The pointy part that sticks out in the image (immediately next to the xyz axis) is a poor-man's chamfer I guess, but it somewhat does the job of having a less steep angle between the base plate and this reducer.

If somebody has some pointers for how to round the `tube to tube-rect` shape that would still be appreciated 🫶

**Comment by adrianVmariano** [comment]
For first shape you could use turtle() to create the path with "arc" to make rounded transitions and then rotate_extrude.   I have the idea that rounding_hole_mask() should be extended to address this case by allowing arbitrary angles instead of just 90 deg.   Another option there might be to use offset_stroke.   I would say that most of the complexity of example 21 of turtle3d() is really parameter management, not the actual geometry itself.  

For the second one, I don't think there's an easy answer.   I would make that transition using skin() instead of hull(), which I think looks better.  See skin() examples 11-12.  You could, if you actually need the flat sections, use the skin() approach to create the transition and then join on flat bits, though it would still have a edge at the transition.
